File: leetcode/Python/0206-e-reverse-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

"""
Linked List
Iterative
"""
class Solution_1:
    def reverseList(self, head: ListNode) -> ListNode:
        prev = None
        curt = head

        while curt:
            temp = curt.next
            curt.next = prev
            prev = curt
            curt = temp

        return prev

"""
Recursive
"""
class Solution_2:
    def reverseList(self, head: ListNode) -> ListNode:
        if not head or not head.next: 
            return head
        new_head = self.reverseList(head.next)
        head.next.next = head
        head.next = None
        return new_head

File: leetcode/C++/0206-e-reverse-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
// Time: O(n)
// Space: O(1)
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        ListNode* prev = nullptr;
        ListNode* cur = head;
        
        while (cur) {
            ListNode* tmp = cur->next;
            cur->next = prev;
            prev = cur;
            cur = tmp;
        }
        return prev;
    }
};


// Time: O(n)
// Space: O(n)
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        if (!head || !head->next)
            return head;
        ListNode* p = reverseList(head->next);
        head->next->next = head;
        head->next = nullptr;
        return p;
    }
};